google/cloud/internal/async_polling_loop.cc

```diff
@@ -133,23 +133,14 @@ class AsyncPollingLoopImpl
     if (op && op->done()) {
       return promise_.set_value(*std::move(op));
     }
-    // Update the polling policy even on successful requests, so we can stop
-    // after too many polling attempts.
+    // Update the polling policy even on successful requests, so we can
+    // cancel the operation after too many polling attempts.
     if (!polling_policy_->OnFailure(op.status())) {
-      if (op) {
-        // We should not be fabricating a `Status` value here. Rather, we
-        // should cancel the operation and wait for the next poll to return
-        // an accurate status to the user, otherwise they will have no idea
-        // how to react.
-        return promise_.set_value(Status(
-            StatusCode::kDeadlineExceeded,
-            location_ + "() - polling loop terminated by polling policy"));
+      if (!op) {
+        return promise_.set_value(std::move(op).status());
       }
-      return promise_.set_value(std::move(op).status());
-    }
-    if (op) {
-      std::unique_lock<std::mutex> lk(mu_);
-      op_name_ = std::move(*op->mutable_name());
+      // Consider trying to reset the backoff wait period.
+      DoCancel();
```

Comment:
There is no guarantee that calling `Cancel()` will actually cancel anything, right?  The `Cancel()` could fail, could be ignored, could take several more minutes to resolve?  This means the operation could be running for much longer than the polling policy prescribes (which sounds bad), including "forever" (which sounds really bad). And there are still cases where you get back an error and don't know what the state of the long running operation is.

Reply:
> There is no guarantee that calling `Cancel()` will actually cancel anything, right? The `Cancel()` could fail, could be ignored, could take several more minutes to resolve? 

Right, although it is "best effort".

> This means the operation could be running for much longer than the polling policy prescribes (which sounds bad), including "forever" (which sounds really bad).

Correct, but this is no worse than before, where we always abandoned the operation, leaving it to run as long as it wanted.  Now, while the poll continues to succeed, we will continue to request cancellation, which is the anti-abandon behavior that I'll claim is a feature.

If a user really wants to abandon an operation, they probably want to use lower-level primitives ala #7658.

> And there are still cases where you get back an error and don't know what the state of the long running operation is.

Yes, the previously mentioned transient-at-exhaustion (which we could think about addressing).

Reply:
> Correct, but this is no worse than before, where we always abandoned the operation, leaving it to run as long as it wanted.

Yes.  I was more worried about the application blocking on `.get()` or never getting a calling in a `.then()`.

> Now, while the poll continues to succeed, we will continue to request cancellation, which is the anti-abandon behavior that I'll claim is a feature.

I counter-claim: returning from `.get()` once the polling policy has expired is a feature too.

So, I like the fact that we don't abandon things, but I don't like the fact that now we are (maybe surprisingly) working -- and maybe blocking -- beyond the constraints in the polling policy. 

Would it be possible to make the new behavior optional?  If not, I think we should leave this alone and give folks this choice (to cancel on a polling expiration) once we implement #7658.

> If a user really wants to abandon an operation, they probably want to use lower-level primitives ala #7658.

I will make some notes there.